Holopad

TermService opened the pull request:
Добавляет в игру голопад, с возможностью связи с другими членами персонала по средствам аудио и видео связи.

![screenshot_3](https://user-images.githubusercontent.com/18397924/46408613-59d55e00-c71b-11e8-96f5-8d13c8989d9c.png)
![hdrfl0m6sfengmufonin8a](https://user-images.githubusercontent.com/18397924/46408657-7ec9d100-c71b-11e8-91af-0e84601c9373.png)



:cl: TermService
- rscadd: добаавлен голопад, с возможностью связи между персоналом.

TermService committed:
add holopad

TermService committed:
style update

TermService committed:
moving to design.dm

TermService committed:
little fix

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,181 @@
+	call_state = CALL_RINGING
+	icon_state = "holopad_ringing"
+	desc = "[initial(desc)] Incoming call from [caller.getName()]"
+	spawn(0)
```
INVOKE_ASYNC

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,181 @@
+
+/obj/item/device/holopad/proc/ring()
+	if(call_state != CALL_RINGING) return
+	if(isliving(loc) && loc:client)
```
Красиво привёл к типу

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,181 @@
+/obj/item/device/holopad/proc/ring()
+	if(call_state != CALL_RINGING) return
+	if(isliving(loc) && loc:client)
+		loc<<"bzzzzt"
```
Красивый вывод в гунчат, или это дебаг проскакивает?

TermService replied:
бля я забыл гунчат, спасибо

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,181 @@
+	if(isliving(loc) && loc:client)
+		loc<<"bzzzzt"
+		playsound(loc, 'sound/machines/twobeep.ogg', 25, -5)
+	spawn(50)
```
addtimer()

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,181 @@
+
+/obj/item/device/holopad/proc/placeCall()
+	var/list/Targets = list()
+	for(var/obj/item/device/holopad/H in world)
```
Я даже не хочу комментировать цикл по всему миру.

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,181 @@
+			abonent.acceptCall()
+			call_state = CALL_IN_CALL
+			icon_state = "holopad_in_call"
+			spawn(1) update_holo()
```
Смотри что делать со спавнами выше

LudwigVonChesterfield reviewed (commented):
Я вообще не всё посмотрел. Там ещё идентичных ошибок куча.

TermService committed:
Добавлены Юлианские спрайты, убраны спауны и заменен to_chat

TermService commented:
![screenshot_14](https://user-images.githubusercontent.com/18397924/48921580-b5b59b00-eeb1-11e8-8d01-b4a42bbcdfc8.png)

Miracler112 commented:
В этот раз всё нормально, ждем повторный ревью Людука, если так же всё будет нормально - то будет готово к мержу

LudwigVonChesterfield commented on code/__HELPERS/icons.dm:
```diff
@@ -886,3 +886,18 @@ var/global/list/humanoid_icon_cache = list()
 	else
 		return humanoid_icon_cache[icon_id]
+
+/proc/build_composite_icon_omnidir(atom/A)
```
У нас точно нету тысячи таких полков до тебя?

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+
+/obj/item/device/holopad/Destroy()
+	if(CALL_IN_CALL)
+		call_state = CALL_NONE
```
Эээ а зачем в дестроере циферку менять

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+		voice = "Holopad [sanitize(abonent.id)]"
+	var/rendered = "<span class='game say'><span class='name'>[voice]</span> transmits, \"<span class='message'>[sanitize(text)]</span>\"</span>"
+	for(var/mob/M in listening)
+		M:show_message(rendered, 2)
```
А зачем тут :?

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+	set name="Set ID"
+	set category = "Object"
+	set src in usr
+	var/newid = input(usr,"What would be new ID?") as text
```
>as text

Но изначально айди то - циферка

TermService replied:
Изначально да, но можно туда вписать имя, должность или хосхуесос

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+
+/obj/item/device/holopad/proc/getName(var/override_busy = 0)
+	if(call_state!=CALL_NONE && !override_busy)
+		return "Holopad [id] - busy"
```
Мне нравиться как везде используется [initial(name)], кроме как тут. А ещё, смешно выйдет если поменять айди во время разговора.

TermService replied:
Ну меняй, энивей звонок будет идти

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+	if(call_state != CALL_RINGING) return
+	var/mob/living/L = loc
+	if(isliving(loc) && L.client)
+		to_chat(loc, "bzzzzt")
```
Пхахвхахаахха

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+	var/list/Targets = list()
+	for(var/obj/item/device/holopad/H in world)
+		if(H == src) continue
+		Targets[H.getName()] = H
```
Мне нравится как у нас может быть два голопада с одним айди

TermService replied:
не вижу проблем в этом

LudwigVonChesterfield replied:
Одному из них нельзя будет дозвониться.

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,189 @@
+	if(!selection) return
+	var/obj/item/device/holopad/target = Targets[selection]
+	if(!target) return
+	if(target.incall(src))
```
Тобишь я без каких-либо проверок могу названивать одному и тому же пользователю, и будет не один гудок, а с десяток?

TermService replied:
нет, во время дозвона холопады перейдут в состояние busy, на оба нельзя устраивать дозвон

LudwigVonChesterfield commented:
А до код стайла докопаться надо?

TermService committed:
mirror update

все таки я думал картинка чуть больше будет выглядеть лучше но нет(в техах текстурка вылазит за тень)

TermService committed:
фикс невозможности дозвона на одинаковые голопады

LudwigVonChesterfield commented on code/__HELPERS/icons.dm:
```diff
@@ -886,3 +886,18 @@ var/global/list/humanoid_icon_cache = list()
+	var/icon/composite = icon('icons/effects/effects.dmi', "icon_state"="nothing")
+	for(var/O in A.underlays)
+		var/image/I = O
+		composite.Blend(new/icon(I.icon, I.icon_state), ICON_OVERLAY)
```
Но это же underlay-и.

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+		return
+	var/mob/living/L = loc
+	if(isliving(loc) && L.client)
+		to_chat(loc, "<span class='warning'>Something vibrates..</span>")
```
Ну имхо написать бы где вибрирует, подумай с этим.

Miracler112 replied:
Если эту вибрацию вообще никто не должен слышать кроме юзера, то зачем здесь лок? В ином случае можно использовать visible_message 🤷‍♂️

TermService replied:
А если он лежит на земле? Никто не увидит что он вибрирует?

LudwigVonChesterfield replied:
Под где - я имел ввиду «в кармане, в сумке», и т.д. Кстати, я тут подумал, мне кажется голопад в сумке работать не будет. У него лок- сумка

TermService replied:
Он же к ливинг приведен

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+
+/obj/item/device/holopad/proc/placeCall()
+	var/list/Targets = list()
+	for(var/obj/item/device/holopad/H)
```
Что? А куда делось in world?

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+		addtimer(CALLBACK(src, .proc/update_holo), 1)
+
+		if(isliving(loc))
+			to_chat(loc, "Connection established")
```
Круто. Может лучше типо Голопад говорит, «установлено подключение»

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+
+/obj/item/device/holopad/atom_init()
+	. = ..()
+	uniq_id = rand(00000, 999999)
```
Я бы ещё тогда проверил есть ли такие же голопады. А то оно не уникально.

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+	id = rand(1000, 9999)
+	name = "[initial(name)] [id]"
+
+/obj/item/device/holopad/Destroy()
```
А зачем переписывать прок если он всего-лишь вызывает родителя.

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+		hologram.pixel_x = ((L.dir&4)?32:((L.dir&8)?-32:0))
+		hologram.pixel_y = ((L.dir&1)?32:((L.dir&2)?-32:0))
+	else if(isturf(loc))
+		hologram.dir = 2
```
Дефайны есть

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+		if(!abonent.hologram)
+			abonent.hologram = new()
+			abonent.hologram.name = "Hologram [sanitize(id)]"
+			abonent.hologram.layer = 5
```
Магия

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+	addtimer(CALLBACK(src, .proc/update_holo_pos), 2)
+
+
+/obj/item/device/holopad/attack_self(mob/user as mob)
```
As mob зачем?

Miracler112 replied:
У нас не используется как приведение : так и преобразование as mob, as turf, as obj и прочее такое, оно тут не нужно

LudwigVonChesterfield commented on code/game/objects/items/devices/holopad.dm:
```diff
@@ -0,0 +1,198 @@
+
+/obj/item/device/holopad/hear_talk(mob/living/M, text, datum/language/speaking)
+	if(call_state == CALL_IN_CALL)
+		abonent.receive(text,M==loc)
```
Мне нравиться как голограммы ещё и переводят с других языков - на нужный.
